### PR TITLE
Adds user agent header for requests to Green Web Foundation APIs

### LINKED ISSUE #181: Add source headers to hosting API calls
**Is your feature request related to a problem? Please describe.**
Not a problem, but something that help us (Green Web Foundation) later down the line to understand where API requests are coming from.

**Describe the solution you'd like**
Add a custom header `x-greencheck-src` to the API fetch requests made to the Greencheck API.

https://github.com/thegreenwebfoundation/co2.js/blob/51b85e23dc1f52ac6642259c09e56fef62948087/src/hosting-api.js#L23-L25

https://github.com/thegreenwebfoundation/co2.js/blob/51b85e23dc1f52ac6642259c09e56fef62948087/src/hosting-api.js#L41

There should be a default value `x-greencheck-src: "co2js"` but there should be a mechanism for that to be changed to a value that's set by the user.

### PR DESCRIPTION
**Pull request by sfishel18**
the version from package.json is used to inject an environment variable when esbuild runs, which the running code can read to create the correct user agent string when making requests.

fixes #181

### CODE REVIEW

**Comment by sfishel18** [comment]
@fershad let me know if this looks like the right direction. there were a couple of things i wasn't sure of:

1. should the requests from the node.js version of the library also include the user agent header? i went ahead and added it there but can easily revert it

2. afaict some of the unit tests in `hosting.test.js` are sending actual network requests. i didn't change that behavior since i'm not sure if it's intentional. but since i added logic to spy on `https.get`, i could easily modify it to mock out the responses if desired.

**Comment by fershad** [comment]
@sfishel18 thanks for this. I'll go through it in the next week or two with a view to merge it into the v0.14.1 release

**Comment by fershad** [comment]
@sfishel18 thanks for this one. I went through it and made some changes:

- I added an optional param (`comment`) to the request headers. This allows a space for users to pass in the name of their app, site, or company which can give us a better insight into where requests are coming from.
- Hitting the actual API is as designed in the tests. I have removed `nock` in some tests & replaced it with spying on the `https.get` request headers.

There's one hurdle for this update though. Right now, the package version gets written to the code when the build is run. However, when we publish to NPM we use the `np` package to smooth out this process.

`np` bumps the package version just before publishing. I believe this should be solvable by using the npm lifecycle hooks which `np` runs (https://github.com/sindresorhus/np?tab=readme-ov-file#npm-hooks). I've added a `version` script to rerun the build when the version number is bumped.

**Comment by fershad** [comment]
@mrchrisadams this PR is ready for a review when you've got a moment.

The changes made add a `User-Agent` header to fetch requests made from CO2.js to the Greencheck API. This will show the current version of CO2.js by default. There's an optional parameter where the user can pass in a string to represent the app, site, or company making the request.

The output of the `check` function remains unchanged.

For example:

```js
import {hosting} from '@tgwf/co2'

hosting.check("google.com", "MyCoolApp").then((result) => {
    console.log(result,);
  });
```

This would send a fetch request to the Greencheck API with the `User-Agent: co2js/0.14.1 MyCoolApp` header.

Note, that when used in the browser the `User-Agent` header will be set to whatever the browser sets since I believe it's read-only in that context.

File: src/hosting-api.js
Comment (mrchrisadams) [inline src/hosting-api.js:10]: Would we want to add the same comment info as used in the other `checkAgainstAPI` function? u.e.

```
* @param {string} comment - Optional. The app, site, or organisation that is making the request.
```

File: src/hosting-api.test.js
Comment (mrchrisadams) [inline src/hosting-api.test.js:13]: AsI understand it, this would move our project from using a _mock_, which doesn't actually make network requests, to using a _spy_,  [which still calls the underlying function](https://jestjs.io/docs/jest-object#jestspyonobject-methodname), resulting in a network call.

If we're already not making network requests, I'd prefer we not get into the habit of making network requests in test suites if possible. However, I get that this is something we can do separately from this PR. If we're not going to add an [mockimplementation](https://jestjs.io/docs/mock-function-api#mockfnmockimplementationfn), in this test, can we make an issue to come back to this, so there's a nice well defined issue for someone who is familiar with jest and mocking to pick up in future?

File: src/hosting.js
Comment (mrchrisadams) [inline src/hosting.js:10]: I think the function signature would need the updated comment here as well. The same comment we use elsewhere ought to work.

```
* @param {string} comment - Optional. The app, site, or organisation that is making the request.
```

File: src/hosting.test.js
Comment (mrchrisadams) [inline src/hosting.test.js:7]: I can't remember, but I think we might no longer need `pagexray` here, as @soulgalore mentioned making a release to move functionality into his library. This would mean we could remove pagexray as discussed it in #182. 

https://github.com/thegreenwebfoundation/co2.js/issues/182

Knowing Peter, I reckon he'd be one the first to add a user agent to requests too, once this feature is merged in 😄 

@fershad is it ok to leave this call with you? It's small change, but it might be betteer to handle outside of this PR.

**Review by mrchrisadams — APPROVED** [review]
Hi @fershad thanks for tagging me - this looks good, and the comments I had are small quibbles, which while being nice to address, would not stop this being merged in.

Feel free to merge in, but do please take a mo to review the comments about whether we update related issues like the pagexray one, or adding the use of jest mocks instead of just spies.

File: src/hosting.test.js
Comment (fershad) [reply src/hosting.test.js:7]: Cheers Chris, we'll remove the `pagexray` stuff in the next minor release of CO2.js so not this update.  I'll create a separate PR to track that bit of work.

> Knowing Peter, I reckon he'd be one the first to add a user agent to requests too, once this feature is merged in 😄

Hahah I thought the same thing ~~

File: src/hosting-api.test.js
Comment (fershad) [reply src/hosting-api.test.js]: Made those changes. Mocked `fetch` and `https.get` API requests.

**Comment by fershad** [comment]
Thanks @mrchrisadams I've made most of those changes, including reverting the API calls to mocks. Going to let this cool for a couple of days before merging & releasing a new build.

@sfishel18 welcome to cast your eye over this as well :)

**Comment by mrchrisadams** [comment]
Thanks for making those changes @fershad particularly adding the networking mocks (I've been feeling a bit uneasy about nock, as it's a library I'm not very familiar with, and was a bit scared of changing it.)

One last thing (and this is likely a response to you saying let's let it cool a couple of days so we can view it in with fresh eyes later) is that we're passing in `comment` as our variable name, but really it's setting a _User-Agent_ string.

It might make sense to just name the argument `userAgent`, `userAgentIdentifier` or something along those lines to make the purpose of this extra argument really explicit - this feels like it might be a bit clearer than `comment` as a variable name in a few months.

I reckon this probably is worth deciding before we merge in, but I'd leave the final decision with you.

WDYT?

**Comment by fershad** [comment]
> It might make sense to just name the argument `userAgent`, `userAgentIdentifier` or something along those lines to make the purpose of this extra argument really explicit  ...

Makes a ton of sense. I've renamed it to `userAgentIdentifier`.

**Comment by sfishel18** [comment]
> @sfishel18 welcome to cast your eye over this as well :)

thanks @fershad! sorry for the delayed reply, just got back from a vacation. all looks good to me. let me know if there's anything in particular still missing here that i could look into.

**Comment by fershad** [comment]
@sfishel18 all good mate, added you to the contributors list. Thanks for helping with this change.

**Comment by sfishel18** [comment]
> @sfishel18 all good mate, added you to the contributors list. Thanks for helping with this change.

my pleasure! let me know if you have any other good starter tasks top-of-mind for someone new. no worries if not, i can look over the open issues and pick one that speaks to me :)

**Pull request merged**
